src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1137 @@
+        key_agreement: CoseKey,
+        pin_auth: Vec<u8>,
+        new_pin_enc: Vec<u8>,
+        pin_hash_enc: Vec<u8>,
+    ) -> Result<(), Ctap2StatusCode> {
+        if persistent_store.pin_retries() == 0 {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_BLOCKED);
+        }
+        let pk: crypto::ecdh::PubKey = CoseKey::try_into(key_agreement)?;
+        let shared_secret = self.key_agreement_key.exchange_x_sha256(&pk);
+
+        let mut auth_param_data = new_pin_enc.clone();
+        auth_param_data.extend(&pin_hash_enc);
+        if !check_pin_auth(&shared_secret, &auth_param_data, &pin_auth) {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID);
+        }
+
+        let aes_enc_key = crypto::aes256::EncryptionKey::new(&shared_secret);
+        let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+        self.check_pin_hash_enc(rng, persistent_store, &aes_dec_key, pin_hash_enc)?;
+
+        if !check_and_store_new_pin(persistent_store, &aes_dec_key, new_pin_enc) {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_POLICY_VIOLATION);
+        }
+        self.pin_uv_auth_token = rng.gen_uniform_u8x32();
+        Ok(())
+    }
+
+    fn process_get_pin_token(
+        &mut self,
+        rng: &mut impl Rng256,
+        persistent_store: &mut PersistentStore,
+        key_agreement: CoseKey,
+        pin_hash_enc: Vec<u8>,
+    ) -> Result<AuthenticatorClientPinResponse, Ctap2StatusCode> {
+        if persistent_store.pin_retries() == 0 {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_BLOCKED);
+        }
+        let pk: crypto::ecdh::PubKey = CoseKey::try_into(key_agreement)?;
+        let shared_secret = self.key_agreement_key.exchange_x_sha256(&pk);
+
+        let aes_enc_key = crypto::aes256::EncryptionKey::new(&shared_secret);
+        let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+        self.check_pin_hash_enc(rng, persistent_store, &aes_dec_key, pin_hash_enc)?;
+
+        // Assuming PIN_TOKEN_LENGTH % block_size == 0 here.
+        let iv = [0u8; 16];
+        let mut blocks = [[0u8; 16]; PIN_TOKEN_LENGTH / 16];
+        for (i, item) in blocks.iter_mut().take(PIN_TOKEN_LENGTH / 16).enumerate() {
+            item.copy_from_slice(&self.pin_uv_auth_token[i * 16..(i + 1) * 16]);
+        }
+        cbc_encrypt(&aes_enc_key, iv, &mut blocks);
+        let mut pin_token = vec![];
+        for item in blocks.iter().take(PIN_TOKEN_LENGTH / 16) {
+            pin_token.extend(item);
+        }
+
+        #[cfg(feature = "with_ctap2_1")]
+        {
+            self.permissions = 0x03;
+            self.permissions_rp_id = None;
+        }
+
+        Ok(AuthenticatorClientPinResponse {
+            key_agreement: None,
+            pin_token: Some(pin_token),
+            retries: None,
+        })
+    }
+
+    #[cfg(feature = "with_ctap2_1")]
+    fn process_get_pin_uv_auth_token_using_uv_with_permissions(
+        &self,
+        _: CoseKey,
```

Comment:
Will this be used in the future PR?

Reply:
Not for the hardware we support right now. We'd need internal user verification like

- fingerprint reader
- PIN pad
- iris scanner ;)

And wanted to include this function with the correct parameters as an entry point in case someone wants to add this feature. GetInfo should already report we don't support internal UV, so we are free to just respond with `CTAP2_ERR_INVALID_*COMMAND` for now.

Reply:
The [specification](https://fidoalliance.org/specs/fido2/fido-client-to-authenticator-protocol-v2.1-rd-20191217.html#gettingPinUvAuthToken) under `Getting pinUvAuthToken using built-in user verification methods` describes this command. It's parameters include `key_agreement`. Anyone who would want to implement internal user verification needs to also implement this function. The new versions of the specification is a bit clearer in describing this subcommand.

In general, the structure of this files follows the naming scheme `process_*`, like `process_get_pin_uv_auth_token_using_uv_with_permissions`. Here, `get_pin_uv_auth_token_using_uv_with_permissions` is the name of a subcommand (however trivial). All of them are mapped with correct parameter set from the specification. I'd find it confusing to have a command that is lacking a parameter only because it is unused so far.

Does that make sense?

Reply:
Then I think the name should at least be `_foo` with a comment explaining why it's currently not used.

Reply:
Updated, renamed and commented.